src/hotspot/cpu/riscv/riscv.ad

```diff
@@ -2444,10 +2451,18 @@ encode %{
     // Recursive lock case
     __ mv(flag, zr);
     __ ld(tmp, Address(disp_hdr, ObjectMonitor::recursions_offset_in_bytes() - markWord::monitor_value));
-    __ add(tmp, tmp, 1u);
+    __ add(tmp, tmp, 1);
     __ sd(tmp, Address(disp_hdr, ObjectMonitor::recursions_offset_in_bytes() - markWord::monitor_value));
 
     __ bind(cont);
+
+    __ bnez(flag, no_count);
+
+    __ ld(tmp, Address(xthread, JavaThread::held_monitor_count_offset()));
```

Comment:
Note that increment/decrement may clobber 't1' register in certain cases which will conflict with the "flag" which aliases 't1' here. It will be safer to use the 'tmp' register here in this context.